### PR TITLE
feat: add system style

### LINKED ISSUE #47299: Use System Default for Initial Theme?
### Describe the Issue

I had an issue with another website that was fixed by clearing my browser cache. As a result, I had to log in to all of my sites again. I ended up on Reddit which defaulted to light mode for unauthenticated users, and realised that freeCodeCamp does the same.

### Affected Page

https://freecodecamp.org

### Steps to Reproduce

1. Log out.
2. See light mode?

### Expected behavior

I'm wondering if we should leverage the user's system setting to determine if we should default to light or dark mode. 

### Screenshots

_No response_

### System

The only relevant thing here is that I'm on Windows, I think.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Sboonny**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Ref #47299

This aims to add the ability to opt in default system mode. by allowing the theme to be null, so it will move to system.

---

This won't close the issue, because I know at least, that I won't create the local storage logic in the PR.

### CODE REVIEW

**Comment by ahmadabdolsaheb** [comment]
Thanks for putting together this pull request. 
Could please provide more context. I am not sure I understood the issue.
Also can't we use one of the current variable sets instead of introducing new ones?

**Comment by Sboonny** [comment]
> Thanks for putting together this pull request. Could please provide more context. I am not sure I understood the issue?

We are using a system, that stops respecting the preferred camper color scheme when they are logged in (although improvement from previous system, it still an issue and bad UX). To respect the camper choices, I had to add variables that change based on the camper system, but stop applying when the Camper changes to dark or light mode. That's caused the creation of `system-variables`.

After creating the system variables, I need a way to make campers opt in the system (like setting the theme value to `null`), so I may change the functionality of the toggle system, haven't applied it yet, because I am getting side tracked of the amount of forms and divs in the setting page.

> can't we use one of the current variable sets instead of introducing new ones

Maybe I have no idea, I have to look into this after I am done refactor the theme and introducing a system button.

**Comment by ojeytonwilliams** [comment]
Just to make sure I understand the objective. This PR should allow the user to switch between three theme states: light, dark and system settings?

**Comment by Sboonny** [comment]
yeab, no local storage stuff and no changing theme while logged out

**Comment by Sboonny** [comment]
> Reading through the original discussion, I thought the consensus was to move away from storing this in the database in favour of local storage.
> 
> Before anything is implemented, there needs to be a clear goal.
> 
> * Where are we storing the data?
> * How many settings do we want? (system until the user clicks then light, dark), (light, dark, system) or something completely different.
> 
> Finally, irrespective of what we chose, there are only two palettes (dark and light), so we don't need more variables.

We have discussed this in issue triage, and Naomi put valid points about storing it in the database, so it gets carried between devices. if someone prefer dark mode and pick dark mode once, they don't have to pick dark mode in every device that camper use.

I am in favor of local storage as I use multiple devices at different times of day at different environment, so I want the system to be respected always, but the issue triage meeting ended with keeping the database as is.

**Comment by Sboonny** [comment]
> Finally, irrespective of what we chose, there are only two palettes (dark and light), so we don't need more variables.

Do you guys know something, I don't? well of course you do, but that not what I mean. I have actually no idea how to do `preferred-color-scheme` without adding more variables. I will try to meet with Ahmad and talk it out. Hopefully I am missing something. 👍

**Comment by ojeytonwilliams** [comment]
> how to do preferred-color-scheme without adding more variables. 

https://github.com/freeCodeCamp/freeCodeCamp/pull/48568 already had that behaviour for logged out users. That can be reused.

Anyways, all I was trying to say is that we need to be clear about what we actually want before writing code.  If there was a consensus reached in the meeting, that's great, but it doesn't seem to have been documented.

From memory, Mrugesh proposed something similar to GitHub's approach i.e. have three settings

1. respect browser/os settings 
2. dark 
3. light

By default (and when the user is logged out) we want 1. Otherwise we want whatever the user chooses via the settings or menu.

That seems fine to me, and builds on https://github.com/freeCodeCamp/freeCodeCamp/issues/47299#issuecomment-1308930719 but I'm curious what @raisedadead, @ahmadabdolsaheb and @naomi-lgbt think.

**Comment by raisedadead** [comment]
Whatever Oliver said is accurate and should be the flow. 

In addition, and **only for signed-in users**, the theme value should be stored/replicated in their DB record and local storage. Any theme change in the settings should happen in this order:

User changes setting --> Update DB --> Update the Local Storage.

The CSS should then use the local storage value as the source of truth for the signed-in user. This will avoid any "flash" of theme change during page navigation and reloads because the CSS would avoid reading the DB values directly.

For non-signed-in users, all the above can go down the drain, and only a simple preference hierarchy can be followed (assuming we let them do this in the menu or something):

1. OS Setting
2. Dark
3. Light

**Comment by ojeytonwilliams** [comment]
Thanks for clarifying the spec @raisedadead. That seems solid.

**Comment by raisedadead** [comment]
We are officially on Node.js 18 now. We have changed the merge requirements accordingly.

You will need to rebase this PR against the main branch OR Click the "Update branch" button to get the new checks going.

Thanks for your contribution.

File: client/src/components/layouts/default.tsx [hunk 1]
```diff
@@ -94,13 +94,6 @@ interface DefaultLayoutProps extends StateProps, DispatchProps {
   t: TFunction;
 }
 
-const getSystemTheme = () =>
-  `${
-    window.matchMedia('(prefers-color-scheme: dark)').matches === true
-      ? 'dark-palette'
-      : 'light-palette'
-  }`;
```
Comment (ojeytonwilliams) [inline client/src/components/layouts/default.tsx:102]: Why was this removed? I thought the point of this was to allow us to detect the user preferences without having to duplicate the css variables a bunch of times.

Comment (Sboonny) [reply client/src/components/layouts/default.tsx:102]: I should have provided content about my meeting with Ahmad, why we decided to keep it in our meeting before holiday. Sorry about it.

## How to solve the issue with JavaScript

If we keep the behavior, then we have to add:

- check if there is `currentTheme`, to apply the system theme
- check after the user logged in, to apply the JavaScript code

Both are doable but not common, but it feels instead of applying the theme to the canvas, you are changing the canvas to fit a theme, which isn't mainstreamed anymore, instead in the mainstream people let the browser provide the user with desired output without hassle.

### Issues with using JavaScript

Although, this's okay for our codebase, because we are using outdated libraries, that don't use `preferred-color-scheme` in them, like in editor, more details in issue #48936.

But if we are going to move forward and adapt more modern and easier solution that allow for us more options, like the UI Component library, that have components (one of them is FormGroup #46747, it always comes to that component 😨) that use `preferred-color-scheme` or other libraries.

Second point is: it's harder to find the issue, when people are used to a different system. Meaning that people who are used to find the theme colors and its styles in one file called `theme.css`, won't expect us to be using JavaScript to set the theme, so it's harder for people who are new and following the standard of the industry to manipulate or fix the themes issues.

---

## But what about issue with variables based on my discussion with Ahmad?

- It will clutter the devtools, which make it harder for developers to manipulate values in the browser.

With the usage of the selectors (if I did it right, I haven't checked because I was focused on the buttons), the variables will be hidden.

Comment (ojeytonwilliams) [reply client/src/components/layouts/default.tsx:102]: As I understand it, the plan was to keep the setting in local storage and update it whenever the user data is fetched. It's not a particularly convoluted flow.

We can put the all themes in `theme.css`, that's fine.  I just don't think we should be duplicating them.

Also, setting the theme via an attribute on the html element is pretty common. Reddit, GitHub and the React docs all do it and they're about as mainstream as it gets.

Comment (Sboonny) [reply client/src/components/layouts/default.tsx:102]: Setting in local storage is related to the navigate bar to change the theme without logging in, i.e. someone wants to change the theme without changing the whole preferred color scheme in their browser.

Using local storage in the setting is counter-intuitive because they are logged in and connected to the database.

---

If you mean that instead of using the values to database, we use the values in the local-storage when changing to system, this will add a can of worms of how to revert to the `preferred-color-scheme`, when the local storage value is filled when they are offline.

---

Yes, using attribute after deciding the desired theme is common, I am not deleting it. I am respecting the browser behavior of allowing it to follow the user preference, which is in GitHub, React docs and Mozilla docs

Picture from GitHub variables, in incognito mode

image

---

> I just don't think we should be duplicating them. 

True, maybe we can remove a selector or two. but I haven't tested (pushed the limits with the selector yet, was focused on apply the opt in system). I may not be able to do it, t.

Comment (ojeytonwilliams) [reply client/src/components/layouts/default.tsx:102]: Fair point. I think GitHub gets away with it by using SASS to define each theme once and then use them repeatedly in the various media queries.

All I'm trying to say is that defining a theme more than once is not a good idea. Even if it doesn't cause weird bugs, it will make maintenance harder.

Comment (Sboonny) [reply client/src/components/layouts/default.tsx:102]: #we-should-use-sassOrtailwind, I do see the point, don't worry I haven't actively tried to refactor it, so there maybe a way to do it in this PR, that doesn't require remaking our theming system that I have in mind

Comment (Sboonny) [reply client/src/components/layouts/default.tsx:102]: this is the minimal selectors that I can think of, it doesn't use `:not` selector in `:root` so the :root is clutter in the devtools instead of the previous way.

Comment (ojeytonwilliams) [reply client/src/components/layouts/default.tsx:102]: That's a neat solution, nice work.  I was worried we'd have to use SASS for a second, there :cold_sweat:

**Comment by bbsmooth** [comment]
Hey, it's me, so you knew I was going to chime in sooner or later.

A few accessibility concerns:

- The three buttons need to be grouped with an accessible name. Usually you would do this with a `fieldset` and `legend` (i.e. the `.setting-theme` div would be a `fieldset` and the `p` would be a `legend`). I understand that `legend`s are notoriously hard to style, so you could use the hidden `legend` trick. Add `aria-hidden="true"` to the `p` element so screen readers won't know it's there. Then add a `legend` above the `p` with the same text as the `p` and add the `sr-only` class to it so that it won't be shown on the page. It's a slight bit of duplicate content but now everyone is happy.
- The alternative option to the above is:
  - add `role="group"` to the `.setting-theme` div 
  - add an `id` to the `p` element
  - add `aria-labelledby="[id of p element]" to the `.setting-theme` div
- The buttons need to use the `aria-pressed` attribute to convey their state. One button will always be `aria-pressed="true"` and the other two will always be `aria-pressed="false"`.
- Do not use the `disabled` attribute on the currently active button. Yes, this means that people will be able to set focus to it with their keyboard or click on it with their mouse. But screen reader users will not see it at all if it is set to `disabled`, so it's a small trade-off for accessibility.

**Comment by Sboonny** [comment]
> Usually you would do this with a `fieldset` and `legend` (i.e. the `.setting-theme` div would be a `fieldset` and the `p` would be a `legend`).

I didn't do so, because I thought the buttons aren't checkboxes or radio buttons. To be honest, I thought I need to change this `"sr-theme": "Theme",` to more accessible wording, something like `change the theme to {{theme}}`.

> I understand that `legend`s are notoriously hard to style

True, maybe I was subconsciously avoiding it 🤔. Doesn't matter anymore, lets implement it 👍.

**Comment by bbsmooth** [comment]
@Sboonny This is getting very close. Just a few things that need to be addressed.

- We should get rid of the "Toggle Checkmark" sr-only text. The `aria-pressed` attribute will tell screen reader users that the button is currently active. This sr-only text is just duplicate, unneeded information for them. The checkmark is only for the benefit of sighted users. Also, the sr-only text is added to the accessible name of the button but it is not really intended to be part of the accessible name. UPDATE: Ahh, I see that all of the toggle buttons on this page have the "Toggle Checkmark" sr-only text. So perhaps this is a wider issue than just this one component. If fixing this would take you out of scope of this PR then don't worry about it now and I will fix it in a separate PR.
- If a user clicks a button that is already pressed then nothing should happen. As it is now, the updated flash message appears on the screen when a currently pressed button is clicked. Again, if fixing this is out of scope then I will fix it in a separate PR.

Also, I'm not sure the `aria-disabled` attribute is needed. I understand why it seems like it would be a good idea, but all of the accessible patterns I have seen for this widget do not use it. I am going to do a little more research and I'll get back to you with my recommendation.

**Comment by ojeytonwilliams** [comment]
Blocked until https://github.com/freeCodeCamp/freeCodeCamp/pull/48974 is in.  Also, per https://github.com/freeCodeCamp/freeCodeCamp/pull/48974#issuecomment-1375515134, we want to add tests to make sure that the editor's theme is in sync with the rest of the site.

**Comment by bbsmooth** [comment]
@Sboonny Concerning `aria-disabled`, after reading up on suggested patterns for exclusive toggle button groups, looking at real world implementations, and getting feedback from accessibility professionals who are much more seasoned than me, it basically comes down to `aria-disabled` is probably not essential here, but it probably doesn't hurt either. 

Ideally, we would do testing with screen reader users to determine whether `aria-disabled` is preferred. I tried to find others who have already done testing on this but couldn't find anything. And the [ARIA Authoring Practices Guide](https://www.w3.org/WAI/ARIA/apg/patterns/) doesn't have anything on this particular pattern. So until we can get some definitive data we are really just guessing.

If I had to pick, I would choose to remove `aria-disabled` from the buttons based on:

- Every exclusive toggle button group implementation I looked at in the "real world" does not use `aria-disabled`. So if this is the most common pattern then screen reader users may be more comfortable with it.
- No articles I found about implementing this pattern used `aria-disabled`, reinforcing that this might be the generally accepted pattern.
- It is generally considered a good practice to use the minimum amount of ARIA needed and since `aria-disabled` basically just mimics `aria-pressed` it could be seen as unnecessary duplicate information.

But really, without testing we can't know for sure which would be preferred (and there is a possibility that testing could give us ambiguous results). So if we want to leave `aria-disabled` on these buttons then you will not get any push back from me (at least not at the moment).

**Review by Sboonny — COMMENTED** [review]
let's delete it 👍

File: client/src/components/layouts/default.tsx (hunk 1 above)
Comment (ahmadabdolsaheb) [reply client/src/components/layouts/default.tsx:102]: Just a heads up that whatever we end up deciding here should be reflected in the Component library's theme system.

**Comment by Sboonny** [comment]
I haven't abandoned this, there is too much part that require changes and I am planning to open issues gradually that reduce the needs for that much change.
